File: src/main/java/com/senderman/lastkatkabot/command/user/Pair.java
package com.senderman.lastkatkabot.command.user;

import com.annimon.tgbotsmodule.api.methods.Methods;
import com.annimon.tgbotsmodule.services.CommonAbsSender;
import com.senderman.lastkatkabot.Love;
import com.senderman.lastkatkabot.command.CommandExecutor;
import com.senderman.lastkatkabot.dbservice.ChatInfoService;
import com.senderman.lastkatkabot.dbservice.ChatUserService;
import com.senderman.lastkatkabot.dbservice.UserStatsService;
import com.senderman.lastkatkabot.model.ChatUser;
import com.senderman.lastkatkabot.service.CurrentTime;
import com.senderman.lastkatkabot.util.Html;
import org.springframework.stereotype.Component;
import org.telegram.telegrambots.meta.api.objects.Message;
import org.telegram.telegrambots.meta.api.objects.User;

import java.util.*;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Future;
import java.util.stream.Collectors;

@Component
public class Pair implements CommandExecutor {

    private final UserStatsService userStats;
    private final ChatUserService chatUsers;
    private final ChatInfoService chatInfoService;
    private final Love love;
    private final CurrentTime currentTime;
    private final Set<Long> runningChatPairsGenerations;
    private final ExecutorService threadPool;

    public Pair(
            UserStatsService userStats,
            ChatUserService chatUsers,
            ChatInfoService chatInfoService,
            Love love,
            CurrentTime currentTime,
            ExecutorService threadPool
    ) {
        this.userStats = userStats;
        this.chatUsers = chatUsers;
        this.chatInfoService = chatInfoService;
        this.love = love;
        this.currentTime = currentTime;
        this.runningChatPairsGenerations = Collections.synchronizedSet(new HashSet<>());
        this.threadPool = threadPool;
    }

    @Override
    public String getTrigger() {
        return "/pair";
    }

    @Override
    public String getDescription() {
        return "пара дня";
    }

    @Override
    public void execute(Message message, CommonAbsSender telegram) {
        var chatId = message.getChatId();

        if (message.isUserMessage()) {
            Methods.sendMessage(chatId, "Команду нельзя использовать в лс!").callAsync(telegram);
            return;
        }

        // check if pair was generated today
        var chatInfo = chatInfoService.findById(chatId);
        List<String> lastPairs = Objects.requireNonNullElseGet(chatInfo.getLastPairs(), ArrayList::new);
        var lastPairGenerationDate = Objects.requireNonNullElse(chatInfo.getLastPairDate(), -1);
        int currentDay = Integer.parseInt(currentTime.getCurrentDay());

        if (!lastPairs.isEmpty() && lastPairGenerationDate == currentDay) {
            Methods.sendMessage(chatId, "Пара дня: " + lastPairs.get(0)).callAsync(telegram);
            return;
        }

        if (runningChatPairsGenerations.contains(chatId)) {
            Methods.sendMessage(chatId, "Пара дня все еще генерируется, подождите!").callAsync(telegram);
            return;
        }

        // clean inactive chat members
        chatUsers.deleteInactiveChatUsers(chatId);

        if (chatUsers.countByChatId(chatId) < 2) {
            Methods.sendMessage(chatId, "Недостаточно пользователей писало в чат за последние 2 недели!").callAsync(telegram);
            return;
        }

        // start chat flooding to make users wait for pair generation
        String[] loveStrings = love.getRandomLoveStrings();
        Future<?> floodFuture = threadPool.submit(() -> sendRandomShitWithDelay(chatId, loveStrings, 1000L, telegram));

        threadPool.execute(() -> {
            runningChatPairsGenerations.add(chatId);
            try {
                PairData pair;
                try {
                    pair = generateNewPair(chatId, telegram);
                } catch (NotEnoughUsersException e) {
                    floodFuture.cancel(true);
                    Methods.sendMessage(chatId,
                            "Ой, а у вас мало пользователей в чате доступно... Придется остановить генерацию пары")
                            .callAsync(telegram);
                    return;
                }

                // save new generated pair and date to DB
                chatInfo.setLastPairDate(currentDay);
                lastPairs.add(0, pair.toString());
                chatInfo.setLastPairs(lastPairs.stream().limit(10).collect(Collectors.toList()));
                chatInfoService.save(chatInfo);

                var text = String.format(loveStrings[loveStrings.length - 1],
                        Html.getUserLink(pair.getFirst()),
                        Html.getUserLink(pair.getSecond()));
                // wait while flood ends to prevent race condition
                floodFuture.get();
                Methods.sendMessage(chatId, text).callAsync(telegram);
            } catch (InterruptedException | ExecutionException e) {
                floodFuture.cancel(true);
                e.printStackTrace();
                Methods.sendMessage(chatId, "...Упс, произошла ошибка. Попробуйте еще раз").callAsync(telegram);
            } finally {
                runningChatPairsGenerations.remove(chatId);
            }
        });
    }

    private PairData generateNewPair(long chatId, CommonAbsSender telegram) throws NotEnoughUsersException {

        final int USERS_REQUIRED = 2;

        List<ChatUser> usersForPair;
        while (true) {
            // request two random users from db
            usersForPair = chatUsers.getTwoOrLessUsersOfChat(chatId);
            if (usersForPair.size() < USERS_REQUIRED)
                throw new NotEnoughUsersException(usersForPair.size(), USERS_REQUIRED);

            var chatUser1 = usersForPair.get(0);
            var chatUser2 = usersForPair.get(1);
            // check that these users are available through getChatMember method
            // if at least one of the users is not available, continue the while loop
            User user1, user2;
            try {
                user1 = getUserFromChatMember(chatId, chatUser1.getUserId(), telegram);
            } catch (NoChatMemberException e) {
                chatUsers.delete(chatUser1);
                continue;
            }
            try {
                user2 = getUserFromChatMember(chatId, chatUser2.getUserId(), telegram);
            } catch (NoChatMemberException e) {
                chatUsers.delete(chatUser2);
                continue;
            }

            // at this point, we can be sure that user1 and user2 are present in chat.
            // now we have to change user2 to user1's lover if needed and if lover available as ChatMember
            boolean isTrueLove = false;
            var user1Stats = userStats.findById(chatUser1.getUserId());
            if (user1Stats.getLoverId() != null) {
                try {
                    user2 = getUserFromChatMember(chatId, user1Stats.getLoverId(), telegram);
                    isTrueLove = true;
                } catch (NoChatMemberException ignored) {
                    // leave as is
                }
            }

            return new PairData(user1, user2, isTrueLove);

        }
    }

    private User getUserFromChatMember(long chatId, int userId, CommonAbsSender telegram) {
        var member = Methods.getChatMember(chatId, userId).call(telegram);
        if (member == null || member.getStatus().equals("left") || member.getStatus().equals("kicked"))
            throw new NoChatMemberException(userId, chatId);
        return member.getUser();
    }

    private void sendRandomShitWithDelay(long chatId, String[] shit, long delay, CommonAbsSender telegram) {
        for (int i = 0; i < shit.length - 1; i++) {
            Methods.sendMessage(chatId, shit[i]).callAsync(telegram);
            try {
                Thread.sleep(delay);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
    }

    private static class PairData {

        private final User first;
        private final User second;
        private final boolean isTrueLove;

        public PairData(User first, User second, boolean isTrueLove) {
            this.first = first;
            this.second = second;
            this.isTrueLove = isTrueLove;
        }

        public User getFirst() {
            return first;
        }

        public User getSecond() {
            return second;
        }

        private String getPairEmoji() {
            return isTrueLove ? "💖" : "❤️";
        }

        @Override
        public String toString() {
            return String.format("%s %s %s",
                    Html.htmlSafe(first.getFirstName()),
                    getPairEmoji(),
                    Html.htmlSafe(second.getFirstName())
            );
        }
    }

    private static class NotEnoughUsersException extends RuntimeException {
        public NotEnoughUsersException(int found, int required) {
            super("Not enough users, found %d, required %d".formatted(found, required));
        }
    }

    private static class NoChatMemberException extends RuntimeException {

        private final int userId;
        private final long chatId;

        public NoChatMemberException(int userId, long chatId) {
            super("No userId %d in chatId %d".formatted(userId, chatId));
            this.userId = userId;
            this.chatId = chatId;
        }

        public int getUserId() {
            return userId;
        }

        public long getChatId() {
            return chatId;
        }
    }

}

File: src/main/kotlin/com/senderman/lastkatkabot/usercommands/Pair.kt
package com.senderman.lastkatkabot.usercommands

import com.senderman.lastkatkabot.LastkatkaBotHandler
import com.senderman.lastkatkabot.Services
import com.senderman.neblib.CommandExecutor
import com.senderman.neblib.TgUser
import org.telegram.telegrambots.meta.api.methods.groupadministration.GetChatMember
import org.telegram.telegrambots.meta.api.objects.Message
import org.telegram.telegrambots.meta.exceptions.TelegramApiException

class Pair(private val handler: LastkatkaBotHandler) : CommandExecutor {
    override val command: String
        get() = "/pair"
    override val desc: String
        get() = "пара дня"

    private data class Lover(val user: TgUser, val isTrueLover: Boolean)


    override fun execute(message: Message) {
        if (message.isUserMessage) return

        val chatId = message.chatId
        // check for existing pair
        if (Services.db.pairExistsToday(chatId)) {
            var pair = Services.db.getPairOfTheDay(chatId)
            pair = "Пара дня: $pair"
            handler.sendMessage(chatId, pair)
            return
        }

        // remove users without activity for 2 weeks and get list of actual users
        Services.db.removeOldUsers(chatId, message.date - 1209600)
        val userIds = Services.db.getChatMembersIds(chatId)

        // generate 2 different random users
        val user1: TgUser
        val user2: TgUser
        val isTrueLove: Boolean
        try {
            user1 = getUserForPair(chatId, userIds)
            userIds.remove(user1.id)
            val lover = getSecondUserForPair(chatId, userIds, user1)
            user2 = lover.user
            isTrueLove = lover.isTrueLover
        } catch (e: Exception) {
            handler.sendMessage(
                chatId,
                "Недостаточно пользователей для создания пары! Подождите, пока кто-то еще напишет в чат!"
            )
            return
        }
        // get a random text and set up a pair
        val loveArray = Services.botConfig.loveStrings
        val loveStrings = loveArray.random().trim().split("\n")
        for (i in 0 until loveStrings.lastIndex) {
            handler.sendMessage(chatId, loveStrings[i])
            Thread.sleep(1500)
        }
        val pair = if (isTrueLove) "${user1.name} \uD83D\uDC96 ${user2.name}" else "${user1.name} ❤ ${user2.name}"
        Services.db.setPair(chatId, pair)
        handler.sendMessage(chatId, java.lang.String.format(loveStrings.last(), user1.link, user2.link))
    }

    private fun getSecondUserForPair(chatId: Long, userIds: MutableList<Int>, first: TgUser): Lover {
        val loverId = Services.db.getLover(first.id)
        return if (loverId in userIds) {
            try {
                Lover(TgUser(handler.execute(GetChatMember().setChatId(chatId).setUserId(loverId)).user), true)
            } catch (e: TelegramApiException) {
                Lover(getUserForPair(chatId, userIds), false)
            }
        } else Lover(getUserForPair(chatId, userIds), false)
    }

    /**
     * @throws Exception if amount of valid users less than 2
     */
    private fun getUserForPair(chatId: Long, userIds: MutableList<Int>): TgUser {
        while (userIds.size > 2) {
            val userId = userIds.random()
            try {
                val member = handler.execute(GetChatMember().setChatId(chatId).setUserId(userId))
                return TgUser(member.user)
            } catch (e: TelegramApiException) {
                Services.db.removeUserFromChat(userId, chatId)
                userIds.remove(userId)
                continue
            }
        }
        throw Exception("Not enough users")
    }
}